// @ant-design/pro-field/es/components/Text/index.js
import "antd/es/input/style";
import _Input from "antd/es/input";
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectSpread from "@babel/runtime/helpers/esm/objectSpread2";
import React, { useEffect, useImperativeHandle, useRef } from 'react';
import { useIntl } from '@ant-design/pro-provider';
/**
 * 最基本的组件，就是个普通的 Input
 *
 * @param
 */

var FieldText = function FieldText(_ref, ref) {
  var text = _ref.text,
      mode = _ref.mode,
      render = _ref.render,
      renderFormItem = _ref.renderFormItem,
      fieldProps = _ref.fieldProps,
      _ref$emptyText = _ref.emptyText,
      emptyText = _ref$emptyText === void 0 ? '-' : _ref$emptyText;
  var intl = useIntl();
  var inputRef = useRef();
  useImperativeHandle(ref, function () {
    return inputRef.current;
  });
  useEffect(function () {
    if (fieldProps.autoFocus) {
      var _inputRef$current;

      inputRef === null || inputRef === void 0 ? void 0 : (_inputRef$current = inputRef.current) === null || _inputRef$current === void 0 ? void 0 : _inputRef$current.focus();
    }
  }, [fieldProps.autoFocus]);

  if (mode === 'read') {
    var dom = text !== null && text !== void 0 ? text : emptyText;

    if (render) {
      var _render;

      return (_render = render(text, _objectSpread({
        mode: mode
      }, fieldProps), /*#__PURE__*/React.createElement(React.Fragment, null, (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.prefix) || '', dom, (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.suffix) || ''))) !== null && _render !== void 0 ? _render : emptyText;
    }

    return /*#__PURE__*/React.createElement(React.Fragment, null, (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.prefix) || '', dom, (fieldProps === null || fieldProps === void 0 ? void 0 : fieldProps.suffix) || '');
  }

  if (mode === 'edit' || mode === 'update') {
    var placeholder = intl.getMessage('tableForm.inputPlaceholder', '请输入');

    var _dom = /*#__PURE__*/React.createElement(_Input, _extends({
      ref: inputRef,
      placeholder: placeholder,
      allowClear: true
    }, fieldProps));

    if (renderFormItem) {
      return renderFormItem(text, _objectSpread({
        mode: mode
      }, fieldProps), _dom);
    }

    return _dom;
  }

  return null;
};

export default /*#__PURE__*/React.forwardRef(FieldText);
