// @ant-design/pro-utils/lib/components/FilterDropdown/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

var _DropdownFooter = _interopRequireDefault(require("../DropdownFooter"));

require("./index.less");

var FilterDropdown = function FilterDropdown(props) {
  var children = props.children,
      label = props.label,
      footer = props.footer,
      disabled = props.disabled,
      onVisibleChange = props.onVisibleChange,
      visible = props.visible,
      footerRender = props.footerRender;

  var _useContext = (0, _react.useContext)(_antd.ConfigProvider.ConfigContext),
      getPrefixCls = _useContext.getPrefixCls;

  var prefixCls = getPrefixCls('pro-core-field-dropdown');
  return /*#__PURE__*/_react.default.createElement(_antd.Dropdown, {
    disabled: disabled,
    trigger: ['click'],
    visible: visible,
    onVisibleChange: onVisibleChange,
    overlay: /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-overlay")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "".concat(prefixCls, "-content")
    }, children), footer && /*#__PURE__*/_react.default.createElement(_DropdownFooter.default, (0, _extends2.default)({
      disabled: disabled,
      footerRender: footerRender
    }, footer)))
  }, /*#__PURE__*/_react.default.createElement("span", {
    className: "".concat(prefixCls, "-label")
  }, label));
};

var _default = FilterDropdown;
exports.default = _default;
